Mathlib/RingTheory/AdicCompletion/AsTensorProduct.lean

```diff
@@ -0,0 +1,150 @@
+/-
+Copyright (c) 2024 Judith Ludwig, Christian Merten. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Judith Ludwig, Christian Merten
+-/
+import Mathlib.LinearAlgebra.TensorProduct.Pi
+import Mathlib.RingTheory.AdicCompletion.Exactness
+import Mathlib.RingTheory.TensorProduct.Basic
+
+/-!
+
+# Adic completion as tensor product
+
+In this file we examine properties of the natural map
+
+`AdicCompletion I R ⊗[R] M →ₗ[AdicCompletion I R] AdicCompletion I M`.
+
+We show
+
+- it is an isomorphism if `M = R^n`.
+- it is surjective, if `M` is a finite `R`-module.
+
+## TODO
+
+- Show that `ofTensorProduct` is an isomorphism for any finite free `R`-module over an arbitrary
+  ring. This is mostly composing with the isomorphism to `R^n` and checking that a diagram commutes.
+
+-/
+
+variable {R : Type*} [CommRing R] (I : Ideal R)
+variable (M : Type*) [AddCommGroup M] [Module R M]
+variable {N : Type*} [AddCommGroup N] [Module R N]
+
+open TensorProduct
+
+namespace AdicCompletion
+
+private
+def ofTensorProductBil : AdicCompletion I R →ₗ[AdicCompletion I R] M →ₗ[R] AdicCompletion I M where
+  toFun r := LinearMap.lsmul (AdicCompletion I R) (AdicCompletion I M) r ∘ₗ of I M
+  map_add' x y := by
+    apply LinearMap.ext
+    simp
+  map_smul' r x := by
+    apply LinearMap.ext
+    intro y
+    simp only [LinearMap.coe_comp, LinearMap.coe_restrictScalars, Function.comp_apply,
+      LinearMap.lsmul_apply, RingHom.id_apply, LinearMap.smul_apply]
+    rw [smul_eq_mul, mul_smul]
+
+@[simp]
+private lemma ofTensorProductBil_apply_apply (r : AdicCompletion I R) (x : M) :
+    ((AdicCompletion.ofTensorProductBil I M) r) x = r • (of I M) x :=
+  rfl
+
+/-- The natural `AdicCompletion I R`-linear map from `AdicCompletion I R ⊗[R] M` to
+the adic completion of `M`. -/
+noncomputable
+def ofTensorProduct : AdicCompletion I R ⊗[R] M →ₗ[AdicCompletion I R] AdicCompletion I M :=
+  TensorProduct.AlgebraTensorModule.lift (ofTensorProductBil I M)
+
+@[simp]
+lemma ofTensorProduct_tmul (r : AdicCompletion I R) (x : M) :
+    ofTensorProduct I M (r ⊗ₜ x) = r • of I M x := by
+  simp [ofTensorProduct]
+
+variable {M} in
+/-- `ofTensorProduct` is functorial in `M`. -/
+lemma ofTensorProduct_naturality (f : M →ₗ[R] N) :
+    map I f ∘ₗ ofTensorProduct I M =
+      ofTensorProduct I N ∘ₗ AlgebraTensorModule.map LinearMap.id f := by
+  ext
+  simp
+
+section PiFintype
+
+/-
+In this section we show that `ofTensorProduct` is an isomorphism if `M = R^n`.
+-/
+
+attribute [-simp] _root_.smul_eq_mul Algebra.id.smul_eq_mul
```

Comment:
```suggestion
```
You can add `-smul_eq_mul, -Algebra.id.smul_eq_mul` in the `simpa` that break below to obtain the same result.